### PR TITLE
Add basic definition of rewrite proof rules

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

File: src/rewriter/rewrite_proof_rule.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ */
+class RewriteProofRule
+{
+ public:
+  RewriteProofRule();
+  /**
+   * Initialize this rule.
+   * @param id The identifier of this rule
+   * @param userFvs The (user-provided) free variable list of the rule. This
+   * is used only to track the original names of the arguments to the rule.
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:44]: what is the use of this?

File: src/rewriter/rewrite_proof_rule.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+   * variables are normalized such that *all* proof rules use the same
+   * variables, per type. In detail, the n^th argument left-to-right of a given
+   * type T is the same for all rules. This is to facilitate parallel matching.
+   * @param cond The conditions of the rule, normalized to fvs.
+   * @param conc The conclusion of the rule, which is an equality of the form
+   * (= t s), where t is specified as rewriting to s. This equality is
+   * normalized to fvs.
+   * @param isFixedPoint Whether the rule should be applied to fixed point in
+   * the strategy
+   * @param isFlatForm Whether the rule is the flat form of the actual rule
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:55]: clarify "flat form"

File: src/rewriter/rewrite_proof_rule.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  const std::vector<Node>& getVarList() const;
+  /** The context that the rule is applied in */
+  Node getContext() const { return d_context; }
+  /** Does this rule have conditions? */
+  bool hasConditions() const;
+  /** Get (declared) conditions */
+  const std::vector<Node>& getConditions() const;
+  /**
+   * Get the conditions in context { vs -> ss }. This may involve running the
+   * side conditions of this method.
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:81]: clarify

File: src/rewriter/rewrite_proof_rule.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  void getMatches(Node h, expr::NotifyMatch* ntm) const;
+  /** Get conclusion of the rule */
+  Node getConclusion() const;
+  /** Get conclusion of the rule for the substituted terms ss */
+  Node getConclusionFor(const std::vector<Node>& ss) const;
+
+  /**
+   * Is variable explicit? An explicit variable is one that does not occur
+   * in a condition and thus its value must be specified in a proof.
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:101]: clarify "must be specified in a proof"

File: src/rewriter/rewrite_proof_rule.h [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  bool getObligations(const std::vector<Node>& vs,
+                      const std::vector<Node>& ss,
+                      std::vector<Node>& vcs) const;
+  /**
+   * Check match, return true if h matches the head of this rule; notifies
+   * the match notify object ntm.
+   *
+   * Note this method is not run as the main matching algorithm for rewrite
+   * proof reconstruction, which considers all rules in parallel.
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:91]: IIUC what you mean is that there is a main algorithm run elsewhere, of which this method is merely a helper?

File: src/rewriter/rewrite_proof_rule.h [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+   * the strategy
+   * @param isFlatForm Whether the rule is the flat form of the actual rule
+   * with the given id.
+   */
+  void init(DslPfRule id,
+            const std::vector<Node>& userFvs,
+            const std::vector<Node>& fvs,
+            const std::vector<Node>& cond,
+            Node conc,
+            Node context,
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:63]: Not documented

File: src/rewriter/rewrite_proof_rule.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  {
+    expr::getFreeVariables(c, fvsCond);
+  }
+  for (const Node& v : fvs)
+  {
+    d_fvs.push_back(v);
+    if (fvsCond.find(v) == fvsCond.end())
+    {
+      d_noOccVars[v] = true;
+    }
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.cpp:78]:
```suggestion
    d_noOccVars[v] = fvsCond.find(v) == fvsCond.end();
```

File: src/rewriter/rewrite_proof_rule.h [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  bool isExplicitVar(Node v) const;
+  /**
+   * Get list context. This returns the parent kind of the list variable v.
+   * For example, for
+   *   (define-rule bool-or-true ((xs Bool :list) (ys Bool :list))
+   *      (or xs true ys) true)
+   * The variable xs has list context `OR`.
+   */
+  Kind getListContext(Node v) const;
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.h:111]: clarify the behavior when v is in an ambiguous context

File: src/rewriter/rewrite_proof_rule.cpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+const std::vector<Node>& RewriteProofRule::getConditions() const
+{
+  return d_cond;
+}
+
+bool RewriteProofRule::getObligations(const std::vector<Node>& vs,
+                                      const std::vector<Node>& ss,
+                                      std::vector<Node>& vcs) const
+{
+  // otherwise, just substitute into each condition
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.cpp:121]: otherwise?

File: src/rewriter/rewrite_proof_rule.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+Node RewriteProofRule::getConclusion() const { return d_conc; }
+
+Node RewriteProofRule::getConclusionFor(const std::vector<Node>& ss) const
+{
+  Assert(d_fvs.size() == ss.size());
+  Node conc = d_conc;
+  if (isFixedPoint())
+  {
+    Node context = getContext();
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.cpp:143]: just use d_context?

File: src/rewriter/rewrite_proof_rule.cpp [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  d_mt.getMatches(h, ntm);
+}
+
+Node RewriteProofRule::getConclusion() const { return d_conc; }
+
+Node RewriteProofRule::getConclusionFor(const std::vector<Node>& ss) const
+{
+  Assert(d_fvs.size() == ss.size());
+  Node conc = d_conc;
+  if (isFixedPoint())
```
Comment (HanielB) [inline src/rewriter/rewrite_proof_rule.cpp:141]: document what is happening in this block

File: src/rewriter/rewrite_proof_rule.h (hunk 5 above)
Comment (ajreynol) [reply src/rewriter/rewrite_proof_rule.h:91]: No, its not a helper for the main algorithm.  It would be a helper in a naive, non-parallel version.

**Comment by ajreynol** [comment]
I've reverted the last change since it impacts behavior.

**Comment by HanielB** [comment]
I see. In this case the way you are using this map is merely as a set. So you should change it to avoid confusion.

File: src/rewriter/rewrite_proof_rule.h (hunk 8 above)
Comment (HanielB) [reply src/rewriter/rewrite_proof_rule.h:118]: Thanks for clarifying. But in that case we should never get to the `return UNDEFINED_KIND;` part of the implementation, right? You could put an assertion there then (or an `Unreachable`).

File: src/rewriter/rewrite_proof_rule.h (hunk 5 above)
Comment (HanielB) [reply src/rewriter/rewrite_proof_rule.h:93]: Could you clarify then this paragraph? I don't get what it's trying to say.

**Comment by ajreynol** [comment]
Yes, although technically there is a performance hit, std::map is faster than std::unordered_set for small collections. Regardless, I don't think this is performance-critical code.

File: src/rewriter/rewrite_proof_rule.h (hunk 8 above)
Comment (ajreynol) [reply src/rewriter/rewrite_proof_rule.h:119]: I believe this is the case where there is no list context, e.g. it is not a list variable. I'll clarify.

**Review by HanielB — APPROVED** [review]
I'm surprised about this difference between map and unordered_set...

**Pull request merged**